expr/parse_filterql.go

```diff
@@ -0,0 +1,343 @@
+package expr
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	u "github.com/araddon/gou"
+	"github.com/araddon/qlbridge/lex"
+)
+
+// Parses Tokens and returns an request.
+func ParseFilterQL(filter string) (*FilterQL, error) {
+	l := lex.NewFilterQLLexer(filter)
+	m := FilterQLParser{l: l, FilterQLTokenPager: NewFilterQLTokenPager(l), buildVm: false}
+	return m.parse()
+}
+func ParseFilterQLVm(filter string) (*FilterQL, error) {
+	l := lex.NewFilterQLLexer(filter)
+	m := FilterQLParser{l: l, FilterQLTokenPager: NewFilterQLTokenPager(l), buildVm: true}
+	return m.parse()
+}
+
+type FilterQL struct {
+	Pos
+	Raw    string       // full original raw statement
+	Filter *Filters     // A top level filter
+	Limit  int          // Limit
+	Offset int          // Offset
+	Alias  string       // Non-Standard sql, alias/name of sql another way of expression Prepared Statement
+	With   u.JsonHelper // Non-Standard SQL for properties/config info, similar to Cassandra with, purse json
+}
+
+func NewFilterQL() *FilterQL {
+	req := &FilterQL{}
+	return req
+}
+
+type Filters struct {
+	Pos
+	Op      lex.TokenType // OR, AND
+	Filters []*FilterExpr
+}
+
+type FilterExpr struct {
+	Pos
+
+	// Exactly one of these will be non-nil
+	Include string   // name of foregin named alias filter to embed
+	Expr    Node     // Node might be nil in which case must have filter
+	Filter  *Filters // might be nil, must have expr
+}
+
+func NewFilters(tok lex.Token) *Filters {
+	return &Filters{Op: tok.T, Filters: make([]*FilterExpr, 0)}
+}
+func NewFilterExpr() *FilterExpr {
+	return &FilterExpr{}
+}
+
+// TokenPager is responsible for determining end of
+// current tree (column, etc)
+type FilterQLTokenPager struct {
+	*LexTokenPager
+	lastKw lex.TokenType
+}
+
+func NewFilterQLTokenPager(lex *lex.Lexer) *FilterQLTokenPager {
+	pager := NewLexTokenPager(lex)
+	return &FilterQLTokenPager{LexTokenPager: pager}
+}
+
+func (m *FilterQLTokenPager) IsEnd() bool {
+	return m.lex.IsEnd()
+}
+func (m *FilterQLTokenPager) ClauseEnd() bool {
+	tok := m.Cur()
+	//u.Debugf("IsEnd()? tok:  %v", tok)
+	switch tok.T {
+	case lex.TokenEOF, lex.TokenEOS, lex.TokenFrom, lex.TokenHaving, lex.TokenComma,
+		lex.TokenIf, lex.TokenAs, lex.TokenLimit, lex.TokenSelect:
+		return true
```

Comment:
simplified it, it is a "tokenpager" that is for recognizing ends of clauses per dialect, generally this "keyword" is part of the lexer itself but our lexer is neutral.